### PR TITLE
Cleanup and Additions

### PR DESCRIPTION
**Pull request by JasperLorelai**
- General changes:
  - MS error files are now named in a date format. An "error date" line was added to the log file.
  - PlaceholderAPIDataSpell - Added variable replacement and internal placeholder support (deep search using `<>` brackets).
  - Added missing DataLivingEntity elements to the DataSpell.
  - Added `spell-damage-types` to InvulnerabilitySpell.
  - Added `hoveringwith (Item)` modifier condition.
  - Added method to add variables externally.
  - Small Cleanup.
- PassiveSpell changes:
  - Added spell debug message to PassiveSpell spell list.
  - Refactored RightClickEntityListener event - PlayerInteractEntityEvent => PlayerInteractAtEntityEvent (it now supports armor stands).
  - Replaced CraftListener's (`craft`) MagicItems with Util's ItemStack reader (predefined item support).
  - Added `magicspellsloaded` passive trigger.
  - Added `gamemodechange (gamemode)` passive trigger.
  - Added `worldchange (world)` passive trigger.
  - Added `spellselect (spells)` passive trigger.
  - Added `inventoryclick (action) (itemOnSlot) (withItem)` passive trigger:
    - Accepts [InventoryAction](https://hub.spigotmc.org/javadocs/bukkit/org/bukkit/event/inventory/InventoryAction.html), or "null" as a string as a wildcard
    - `itemOnSlot` is the item currently on the slot as the event fired. This can be "null" as a string to represent a wildcard.
    - `withItem` is the item currently hovered as the event fired. It doesn't need to be specifed.
    - None of these parameters have to be specified if you need a catch-all listener for any click, any item, etc.
- RotateSpell changes:
  - Removed `face-target` and `face-caster`.
  - Added `face` (String), can be: `caster`, `target`, `away-from-caster` or `away-from-target.
  - Replaced `rotation` with `rotation-yaw`.
- UnconjureSpell changes:
  - The spell can now erase items hovered in an open inventory.
  - Every listed item accepts a quantity parameter seperated by a space from the item.
  - If the quantity is not specified the behaviour is kept - all items matched (and quantity ignored) are removed.
  - If the quantity is specified the spell will only delete the stack of items that matches it.
  - The sequence of the search is: hovered item, inventory contents, armor contents.
- MS Variable changes:
  - Major rework of variable math - made it so all math is done in one place (VariableMod class).
  - Added `meta_absorption` (get & set).
  - Fixed power operation - `^2`.
  - Added modulo operation - `%10`.
  - Added randomiser operation - `?10` - where 10 defines bounds (So this will return 0-9).
  - All variable operations respect `min` and `max` bounds.
- Spell Effects:
  - Added `category` to `sound` and `soundpersonal` effects.
  - Added `location-modifiers`.
  - Modifiers now control if the effect should play or not.
  - Do note that there are no `target-modifiers`, only `modifiers` for effects played on entities and `location-modifiers` for effects played at locations.

### CODE REVIEW

File: core/src/main/java/com/nisovin/magicspells/variables/VariableManager.java [hunk 1]
```diff
@@ -157,46 +148,46 @@ public void set(String variable, String player, String amount) {
...
 	public Variable getVariable(String name) {
 		return variables.get(name);
 	}
-	
+
 	public void reset(String variable, Player player) {
 		Variable var = variables.get(variable);
 		if (var == null) return;
 		var.reset(player);
-		updateBossBar(var, player != null ? player.getName() : "");
```
Comment (TheComputerGeek2) [inline core/src/main/java/com/nisovin/magicspells/variables/VariableManager.java]: Why removing the null checks?

File: core/src/main/java/com/nisovin/magicspells/variables/VariableManager.java [hunk 2]
```diff
@@ -99,6 +99,16 @@ public VariableManager(MagicSpells plugin, ConfigurationSection section) {
...
 		}, TimeUtil.TICKS_PER_MINUTE, TimeUtil.TICKS_PER_MINUTE);
 	}
+
+	/**
+	 * Adds a variable with the provided name to the list of variables.
+	 * This will replace existing variables if the same name is used.
+	 * @param name the name of the variable
+	 * @param variable the variable to add
+	 */
+	public void addVariable(String name, Variable variable) {
```
Comment (TheComputerGeek2) [inline core/src/main/java/com/nisovin/magicspells/variables/VariableManager.java]: Please provide some indication of if an override occurred

File: core/src/main/java/com/nisovin/magicspells/util/VariableMod.java [hunk 3]
```diff
@@ -12,19 +12,19 @@
...
 		
 		SET((a, b) -> b),
-		ADD((a, b) -> a + b),
+		ADD(Double::sum),
 		MULTIPLY((a, b) -> a * b),
 		DIVIDE((a, b) -> a / b),
-		POWER((a, b) -> FastMath.pow(a, b));
+		MODULO((a, b) -> a % b),
+		POWER(FastMath::pow),
+		RANDOM((a, b) -> Double.valueOf(Util.getRandomInt(b.intValue())));
```
Comment (TheComputerGeek2) [inline core/src/main/java/com/nisovin/magicspells/util/VariableMod.java]: How was the decision reached to lock it to integer values?

Comment (JasperLorelai) [reply core/src/main/java/com/nisovin/magicspells/util/VariableMod.java]: With lack of experience...

**Pull request merged**